src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -400,6 +422,138 @@ protected Object matchInRegion(
         }
     }
 
+    @Primitive(name = "regexp_match_in_region_tregex", lowerFixnum = { 2, 3, 5 })
+    public abstract static class MatchInRegionTRegexNode extends PrimitiveArrayArgumentsNode {
+
+        @Child MatchInRegionNode fallbackMatchInRegionNode;
+        @Child DispatchNode warnOnFallbackNode;
+
+        @Child DispatchNode stringDupNode;
+        @Child DispatchNode tRegexGroupCountNode;
+        @Child DispatchNode tRegexGetStartNode;
+        @Child DispatchNode tRegexGetEndNode;
+
+        @Specialization(guards = "libString.isRubyString(string)")
+        protected Object matchInRegionTRegex(
+                RubyRegexp regexp, Object string, int fromPos, int toPos, boolean atStart, int startPos,
+                @Cached ConditionProfile matchFoundProfile,
+                @Cached ConditionProfile tRegexCouldNotCompileProfile,
+                @Cached ConditionProfile tRegexIncompatibleProfile,
+                @Cached LoopConditionProfile loopProfile,
+                @Cached DispatchNode tRegexExecBytesNode,
+                @Cached DispatchNode tRegexIsMatchNode,
+                @Cached RopeNodes.BytesNode bytesNode,
+                @Cached MatchDataNodes.MatchDataCreateNode matchDataCreateNode,
+                @Cached SelectEncodingNode selectEncodingNode,
+                @Cached TRegexCompileNode tRegexCompileNode,
+                @CachedLibrary(limit = "2") RubyStringLibrary libString) {
+            final Rope rope = libString.getRope(string);
+
+            if (tRegexIncompatibleProfile
+                    .profile(toPos < fromPos || toPos != rope.byteLength() || startPos != 0 || fromPos < 0)) {
+                return fallbackToJoni(regexp, string, fromPos, toPos, atStart, startPos);
+            } else {
+                final RubyEncoding encoding = selectEncodingNode.executeSelectEncoding(regexp, string);
+                final Object compiledRegex = tRegexCompileNode.executeTRegexCompile(regexp, atStart, encoding);
+
+                if (tRegexCouldNotCompileProfile.profile(compiledRegex == nil)) {
+                    return fallbackToJoni(regexp, string, fromPos, toPos, atStart, startPos);
+                }
+
+                final byte[] bytes = bytesNode.execute(rope);
+                final Object regexResult = tRegexExecBytesNode
+                        .call(compiledRegex, "execBytes", getContext().getEnv().asGuestValue(bytes), fromPos);
```

Comment:
This and other "foreign calls" should be an InteropLibrary#invokeMember() directly to avoid quite a few extra nodes in between.

Reply:
A little bit trickier than I thought, because `isMatch` and `groupCount` are actually readMember, but easy enough.